torch/csrc/autograd/engine.cpp

```diff
@@ -934,24 +934,27 @@ std::shared_ptr<FutureVariableList> Engine::execute_with_graph_task(
 }
 
 void Engine::mark_graph_task_completed(const std::shared_ptr<GraphTask>& graph_task) {
-  if (graph_task->future_completed_.load()) {
+  // Allow only one thread one attempt to process this logic.
+  if (graph_task->future_completed_.exchange(true) ||
+      graph_task->future_result_->completed()) {
```

Comment:
How can the future_result_ be completed without the future_completed_ flag to be already set?

Reply:
btw, I also expect that this second line isn't necessary.

Reply:
Good point, I'll remove the extra check.